component/publisher-client/src/main/java/org/wso2/am/analytics/publisher/client/EventHubClient.java

```diff
@@ -32,74 +45,218 @@
-            log.error("EventHubClient initialization failed.");
-            return;
+    public EventHubClient(String authEndpoint, String authToken, AmqpRetryOptions retryOptions) {
+        threadBarrier = new ReentrantLock();
+        waitCondition = threadBarrier.newCondition();
+        readWriteLock = new ReentrantReadWriteLock(true);
+        scheduledExecutorService = Executors.newScheduledThreadPool(2, new DefaultAnalyticsThreadFactory(
+                "Reconnection-Service"));
+        producerRetryCounter = new BackoffRetryCounter();
+        eventBatchRetryCounter = new BackoffRetryCounter();
+        this.authEndpoint = authEndpoint;
+        this.authToken = authToken;
+        this.retryOptions = retryOptions;
+        createProducerWithRetry(authEndpoint, authToken, retryOptions, true);
+    }
+
+    private void retryWithBackoff(String authEndpoint, String authToken,
+                                  AmqpRetryOptions retryOptions, boolean createBatch) {
+        scheduledExecutorService.schedule(new Runnable() {
+            @Override
+            public void run() {
+                createProducerWithRetry(authEndpoint, authToken, retryOptions, createBatch);
+            }
+        }, producerRetryCounter.getTimeIntervalMillis(), TimeUnit.MILLISECONDS);
+        producerRetryCounter.increment();
+    }
+
+    private void createProducerWithRetry(String authEndpoint, String authToken,
+                                         AmqpRetryOptions retryOptions, boolean createBatch) {
+        try {
+            if (producer != null) {
+                producer.close();
+            }
+            producer = EventHubProducerClientFactory.create(authEndpoint, authToken, retryOptions);
+            try {
+                if (createBatch) {
+                    batch = producer.createBatch();
+                }
+            } catch (IllegalStateException e) {
+                throw new ConnectionRecoverableException("Event batch creation failed. " + e.getMessage()
+                        .replaceAll("[\r\n]", ""));
+            }
+            clientStatus = ClientStatus.CONNECTED;
+            producerRetryCounter.reset();
+            try {
+                threadBarrier.lock();
+                waitCondition.signalAll();
+            } finally {
+                threadBarrier.unlock();
+            }
+        } catch (ConnectionRecoverableException e) {
+            clientStatus = ClientStatus.RETRYING;
+            log.error("Recoverable error occurred when creating Eventhub Client. Retry attempts will be made. Reason :"
+                              + e.getMessage().replaceAll("[\r\n]", ""));
+            log.debug("Recoverable error occurred when creating Eventhub Client using following attributes. Auth "
+                              + "endpoint: " + authEndpoint.replaceAll("[\r\n]", "") + ". Retry attempts will be made. "
+                              + "Reason : " + e.getMessage().replaceAll("[\r\n]", ""), e);
```

Comment:
any reason to log the same error in a different format and level?

Reply:
Auth endpoint is printed in the debug log. It wont be visible in the normal info log